### PR TITLE
feat: add autoFillPasswords capability

### PR DESCRIPTION
**Pull request by rerorero**
This PR adds `autoFillPasswords` capability that allows to switch AutoFill Passwords settings for a simulator.
Related PR: https://github.com/appium/appium-ios-simulator/pull/378

Looks like prettier modifies a lot of lines in pre-commit hook. I guess it's due to this recent change?
https://prettier.io/docs/en/options.html#trailing-commas
> Default value changed from es5 to all in v3.0.0

Should I keep current trailing comma rule by adding prettierc?

### CODE REVIEW

File: lib/desired-caps.js [hunk 1]
```diff
@@ -2,138 +2,139 @@
 const PLATFORM_NAME_IOS = 'iOS';
 const PLATFORM_NAME_TVOS = 'tvOS';
 
-const desiredCapConstraints = /** @type {const} */({
-  platformName: { // override
+const desiredCapConstraints = /** @type {const} */ ({
```
Comment (mykola-mokhnach) [inline lib/desired-caps.js:5]: Please revert autoformatter changes. They make it very hard to distinguish where the "real" difference is

File: lib/driver.js [hunk 2]
```diff
@@ -622,6 +622,12 @@ class XCUITestDriver extends BaseDriver {
         await this.opts.device.setReduceTransparency(this.opts.reduceTransparency);
       }
 
+      if (_.isBoolean(this.opts.autoFillPasswords)) {
```
Comment (mykola-mokhnach) [inline lib/driver.js:625]: I think this whole settings check block is too long and should be extracted to a separated method to improve the readability

File: lib/desired-caps.js (hunk 1 above)
Comment (rerorero) [reply lib/desired-caps.js:5]: Yeah, currently the original files are not aligned with prettier rules but this repository forces developers to format in each commit with husky.
Can I submit a PR to apply the format to all files?

Comment (rerorero) [reply lib/desired-caps.js:5]: Can you take a look? https://github.com/appium/appium-xcuitest-driver/pull/1973

**Comment by mykola-mokhnach** [comment]
would it be possible to rebase this PR?

**Comment by rerorero** [comment]
Looks like the the test I have added didn't pass but the status is green 🤔 I will check it.
https://pipelinesghubeus2.actions.githubusercontent.com/gIP2FJHApXY0M4FInFE0NsrKdj6XQtvFvsowyqrKwMWjJ9ZUQ3/_apis/pipelines/1/runs/16402/signedlogcontent/2?urlExpires=2023-09-11T09%3A55%3A38.1987357Z&urlSigningMethod=HMACV1&urlSignature=sEf%2BEWVieBtPM56GjChrsnoIsg7s38loHqNMEZSX8Ic%3D

**Comment by rerorero** [comment]
> but the status is green

[Ah it's because of threshold](https://github.com/appium/appium-xcuitest-driver/blob/218bfec69dca957578bb9e49c23e3e071fa94b74/.github/workflows/functional-test.yml#L122)
anyway I'll try to fix my tests

File: test/functional/device/passwords-e2e-specs.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  describe('AutoFillPasswords', function () {
+    async function isPasswordsMenuShown(driver) {
+      const el = await driver.$('~Text Fields');
+      await el.click();
+
+      const textField = await driver.$('XCUIElementTypeSecureTextField');
+      await textField.waitForExist({timeout: 500});
+      await textField.click();
+
+      const passwordsMenu = await driver.$('~Passwords');
```
Comment (rerorero) [inline test/functional/device/passwords-e2e-specs.js:40]: Sorry, I realized this way doesn't work well with iOS 16.2 & iOS 15.5. I only checked with iOS 16.4 & iOS 17.0 when I was [working on this](https://github.com/appium/appium-ios-simulator/pull/378). I tested if the autofill is enabled by checking if the keyboard has a Password menu like this
スクリーンショット 2023-09-11 19 53 34

Looks like this feature is enabled not only with AutoFill but also Keychain or any other password manager being selected. It doesn't appear if none of the password managers is selected. And, while Keychain is enabled by default in iOS 16.4, it's disabled by default in iOS 16.2.
iOS 16.4:
スクリーンショット 2023-09-11 19 52 44
iOS 16.2:
スクリーンショット 2023-09-11 19 49 12

Also `setAutoFillPasswords` doesn't work well with iOS 15.5 for another reason. So basically this only works with iOS 16.4+ for now, as long as the keychain setting keeps default. Actually, this is enough for my issue: I just want to disable it in iOS 16.4 to align the default behavior with older OS like iOS 15. But now I'm feeling bad about making this available in this driver as it is. I think I should look for a better way by having a further look at other related plist files that I mentioned or Private Frameworks interfaces.

Comment (mykola-mokhnach) [reply test/functional/device/passwords-e2e-specs.js:40]: The current implementation seems ok for me as soon as we properly describe the requirements in README
Additional features may always be added later
What do you think @KazuCocoa ?

File: lib/driver.js [hunk 4]
```diff
@@ -711,6 +660,70 @@ class XCUITestDriver extends BaseDriver {
     }
   }
 
+  /**
+   * Start the simulator and initialize based on capabilities
+   */
+  async startAndSetupSimulator() {
```
Comment (mykola-mokhnach) [inline lib/driver.js:666]: initSimulator would be shorter

File: lib/driver.js [hunk 5]
```diff
@@ -711,6 +660,70 @@ class XCUITestDriver extends BaseDriver {
...
+    if (await setSafariPrefs(this.opts.device, this.opts)) {
+      this.log.debug('Safari preferences have been updated');
+    }
+
+    // @ts-expect-error - do not assign arbitrary properties to `this.opts`
+    if (await setLocalizationPrefs(this.opts.device, this.opts)) {
+      this.log.debug('Localization preferences have been updated');
+    }
+
+    if (_.isBoolean(this.opts.reduceMotion)) {
```
Comment (mykola-mokhnach) [inline lib/driver.js:692]: the next three blocks could be merged  to a loop:

```
const promises = [];
for (const optName of ['reduceMotion', 'reduceTransparency', 'autoFillPasswords']) {
  promises.push(device[`set${_.toUpper(optName[0])}${optName.substring(1)}`]());
}
await B.all(promises);
```

File: test/functional/device/passwords-e2e-specs.js (hunk 3 above)
Comment (KazuCocoa) [reply test/functional/device/passwords-e2e-specs.js:40]: Agree. Addressing the limitation in the README should be sufficient in the case i believe

Comment (rerorero) [reply test/functional/device/passwords-e2e-specs.js:40]: I got it, will add limitations to README. Thank you.

**Comment by rerorero** [comment]
Thanks, I made changes based on your comments, PTAL.

**Review by KazuCocoa — APPROVED** [review]
lgtm

File: lib/driver.js [hunk 6]
```diff
@@ -711,6 +660,56 @@ class XCUITestDriver extends BaseDriver {
...
+      this.ensureFeatureEnabled(SHUTDOWN_OTHER_FEAT_NAME);
+      await shutdownOtherSimulators(device);
+    }
+
+    await this.startSim();
+
+    if (this.opts.customSSLCert) {
+      // Simulator must be booted in order to call this helper
+      // @ts-expect-error - do not assign arbitrary properties to `this.opts`
+      await this.opts.device.addCertificate(this.opts.customSSLCert);
```
Comment (mykola-mokhnach) [inline lib/driver.js:680]: this.opts.device -> device

**Review by mykola-mokhnach — APPROVED** [review]

File: lib/driver.js (hunk 6 above)
Comment (rerorero) [reply lib/driver.js]: oops, fixed https://github.com/appium/appium-xcuitest-driver/pull/1972/commits/2e6a3c4e87d1f3787af67e4d0bcdfcfd4f178c0b

thank you for reviewing

**Comment by KazuCocoa** [comment]
hm, auto release is not working. let me check

**Pull request merged**